### DISCUSSION TITLE
Embedded Nats Server nuget package

### DISCUSSION THREAD

**Discussion by niklasfp** [nats-io/nats.net.v1#760] [Ideas]
Hi I'm currently experimenting with creating an embedded nats server for dotnet via a multiplatform nuget package that contains native binaries for the nats server (linux amd64 and winx64 for now)

I'm sure that this is not something for this repo, since its a client and pretty slim atm, but there could be synergies going forward, especially if find a way to use the client via a in process pipe, sort of like how the go client is able to connect directly.

So why am I posting here, mostly to hear if this is something that others have a need for, and if that's the case I'll make the stuff public and hopefully have some more experienced nats / go guys help out.

What I have so far is a small go package that uses the nats server package and exposes some simple methods, that in turn is compiled to a.so and.dll utilizing go's cgo capabilities (go builmode=c-shared) and packed in a multiplatform dotnet nuget package.

This let's me have nats-server in-process wrapped in a Nats Server class to hide away the ffi part :)
NOTE: For now I've just exposed enough to:
1. create a server with config (accepting same args as nats-server) and add a custom logger to nats in go that forwards logs via a callback
2. Start the server
3. Stop the server

This is an extract from the generated c header
```c
extern uintptr_t CreateServer(char* arguments, loggerFunc loggerFuncPtr);
extern GoInt StartServer(uintptr_t handle);
extern GoInt ShutdownServer(uintptr_t handle);
extern void FreeServer(uintptr_t handle);

```

For now I use the nats client for all management etc so still going over a socket, but loopback

This is an example on how I use my Nats Server class  in a console test app in c#

```csharp
var services = new ServiceCollection()
    .AddLogging(c =>
        c.AddConsole())
    .AddSingleton<NatsServer.Net.NatsServer>()
    .AddSingleton(new NatsServerOptions()
        {
            Arguments = new[]
            {
                "-js",
            }
        })
        .BuildServiceProvider();

var server = services.GetRequiredService<NatsServer.Net.NatsServer>();

var cts = new CancellationTokenSource();
server.Start(cts.Token);

**Comment by scottf** [comment]
Have you looked at the integration tests in the client? We run one or more servers from code for each test suite. I do think that this would be extremely useful in it's own repo that other projects could access via nuget. We have the exact same idea for java (https://github.com/nats-io/java-nats-server-runner)

**Reply by niklasfp** [reply]
@scottf the java runner, it's "just" a wrapper that runs the nats server go exe right?

Btw you can check it out here:
https://github.com/niklasfp/dotnet-nats-server

The stuff that embeds the server in a library is in the go folder and the dotnet stuff in the dotnet folder.

A few notes:
 I'm not a go programmer, so there might be weird stuff there (it works though 😂)
There are no published nuget packages yet
Ci and Cross compiling is not yet set up, but works for me on both linux and windows.

It's early days for that repo, up until now more or less an experimemt

**Comment by scottf** [comment]
The java starts an external process. Same with the .net one. Take a look at this: https://github.com/nats-io/nats.net/blob/master/src/Tests/IntegrationTests/TestUtilities.cs

I think the best thing is for you to build your project however you want in your github org and we can add it to this page: https://nats.io/download/#connectors-and-utilities
